src/Altinn.ApiClients.Maskinporten/Service/MaskinportenService.cs

```diff
@@ -0,0 +1,155 @@
+            _client = httpClient;
+            _maskinportenConfig = maskinportenConfig.Value;
+        }
+
+     
+        public async Task<string> GetToken(X509Certificate2 cert, string clientId, string scope, string resource, int exp)
+        {
+            return await GetToken(cert, null, clientId, scope, resource, exp);
+        }
+
+        public async Task<string> GetToken(JsonWebKey jwk, string clientId, string scope, string resource, int exp)
+        {
+            return await GetToken(null, jwk, clientId, scope, resource, exp);
+        }
+
+        public async Task<string> GetToken(string base64EncodedJwk, string clientId, string scope, string resource, int exp)
+        {
+            byte[] base64EncodedBytes = Convert.FromBase64String(base64EncodedJwk);
+            string jwkjson = Encoding.UTF8.GetString(base64EncodedBytes);
+            JsonWebKey jwk = new JsonWebKey(jwkjson);
+            return await GetToken(null, jwk, clientId, scope, resource, exp);
+        }
+
+
+        private async Task<string> GetToken(X509Certificate2 cert, JsonWebKey jwk, string clientId, string scope, string resource, int exp)
+        {
+            string jwtAssertion = GetJwtAssertion(cert, jwk, clientId, scope, resource, exp);
+            FormUrlEncodedContent content = GetUrlEncodedContent(jwtAssertion);
+            string maskinPortenToken = await PostToken(content);
+
+            if (!string.IsNullOrEmpty(maskinPortenToken))
+            {
+                var accessTokenObject = JsonConvert.DeserializeObject<JObject>(maskinPortenToken);
+                string accessToken = accessTokenObject.GetValue("access_token").ToString();
+                return accessToken;
+            }
+
+            return null;
+        }
+
+        public string GetJwtAssertion(X509Certificate2 cert, JsonWebKey jwk, string clientId, string scope, string resource, int exp)
+        {
+            DateTimeOffset dateTimeOffset = new DateTimeOffset(DateTime.UtcNow);
+            JwtHeader header = GetHeader(cert, jwk);
+
+            JwtPayload payload = new JwtPayload
+            {
+                { "aud", _maskinportenConfig.JwtAssertionAudience },
+                { "scope", scope },
+                { "iss", clientId },
+                { "exp", dateTimeOffset.ToUnixTimeSeconds() + 100 },
+                { "iat", dateTimeOffset.ToUnixTimeSeconds() },
+                { "jti", Guid.NewGuid().ToString() },
+            };
+            
+            if (string.IsNullOrEmpty(resource))
+            {
+                payload.Add("resource", resource);
+            }
+
+            JwtSecurityToken securityToken = new JwtSecurityToken(header, payload);
+            JwtSecurityTokenHandler handler = new JwtSecurityTokenHandler();
+
+            return handler.WriteToken(securityToken);
+        }
+
+        private JwtHeader GetHeader(X509Certificate2 cert, JsonWebKey jwk)
```

Comment:
This currently feel like two methods in one. There is nothing in common between the two branches.

Reply:
Identifying correct branch/method is something I consider the job of the callers.

Reply:
Fixed